solarforecastarbiter/reference_forecasts/main.py

```diff
@@ -473,19 +477,75 @@ def make_latest_nwp_forecasts(token, run_time, issue_buffer, base_url=None):
     base_url : str or None, default None
         Alternate base_url of the API
     """
+    session, forecast_df = _get_nwp_forecast_df(token, run_time, base_url)
+    execute_for = forecast_df[
+        forecast_df.next_issue_time <= run_time + issue_buffer]
+    if execute_for.empty:
+        logger.info('No forecasts to be made at %s', run_time)
+        return
+    process_nwp_forecast_groups(session, run_time, execute_for)
+
+
+def _get_nwp_forecast_df(token, run_time, base_url):
     session = api.APISession(token, base_url=base_url)
     user_info = session.get_user_info()
     forecasts = session.list_forecasts()
     forecasts += session.list_probabilistic_forecasts()
     forecasts = [fx for fx in forecasts
                  if fx.provider == user_info['organization']]
     forecast_df = find_reference_nwp_forecasts(forecasts, run_time)
-    execute_for = forecast_df[
-        forecast_df.next_issue_time <= run_time + issue_buffer]
-    if execute_for.empty:
-        logger.info('No forecasts to be made at %s', run_time)
-        return
-    process_nwp_forecast_groups(session, run_time, execute_for)
+    return session, forecast_df
+
+
+def _nwp_issue_time_generator(fx, last_timestamp, next_timestamp):
+    # time that will produce forecast value at next_timestamp
+    max_run_time = utils.find_next_issue_time_from_last_forecast(
+        fx, next_timestamp - pd.Timedelta('1ns'))
+    # time that will make forecast for after last_timestamp
+    next_issue_time = utils.find_next_issue_time_from_last_forecast(
+            fx, last_timestamp)
+    while next_issue_time < max_run_time:
+        yield next_issue_time
+        next_issue_time = utils.get_next_issue_time(
+            fx, next_issue_time + pd.Timedelta('1ns'))
+
+
+def _find_group_gaps(session, forecasts, start, end):
+    times = set()
```

Comment:
if you can trust the order from the generator, which I think you can, you can use `dict.fromkeys` as done here: https://github.com/SolarArbiter/solarforecastarbiter-core/blob/839f4ecc61b52fa43fedc29c3020f0339eac78e9/solarforecastarbiter/metrics/preprocessing.py#L293-L298 Not clear to me that one is more readable than the other. I'm sure performance doesn't matter for this.

Reply:
for a single forecast, I think that would be fine. But for multiple forecasts that may have different gaps, I think sticking with the explicit sort is safer and possibly required